### ISSUE TITLE
Implement support for merging nested lambda's

### ISSUE THREAD

**Issue by heckej** [heckej/proguard#1]
# Problem

The lambda's in the following code are not always merged correctly by the simple merger:

```
fun main() {
  val outerLambda = {
    val innerLambda = {
      println("Hello!")
    }
    innerLambda()
  outerLambda()
}
```
In case the order of adding the lambda's to the new lambda group is

1. `innerLambda`
2. `outerLambda`

then the merging succeeds. However, if the order is reversed:

1. `outerLambda`
2. `innerLambda`

then merging the `innerLambda` will result in modifications of the `outerLambda`, but these modifications are not added to the lambda group, so the resulting lambda group is incorrect. More precisely:

1. the implementation of `outerLambda` contains a reference to the original lambda class of `innerLambda`, which is removed by the merger
2. the implementation of `outerLambda` does not reference the merged implementation of `innerLambda` inside the lambda group, so the program behaviour has been altered

# Proposed solution

Make the lambda merger aware of the fact that lambda's can be nested. 

* A correct, but undesirable solution would be to to avoid merging nested lambda's at all.
* The preferred solution is to let the merger merge the inner lambda's before trying to merge the outer lambda
  * This can be done recursively
  * or using a queue of lambda's that still have to be merged, to which outer lambda's are added if they are encountered before the inner lambda's have been merged

# Requirements

* A mechanism to check whether a lambda contains nested lambda's
* A mechanism to check whether a lambda has already been merged

# Testing

* Assert that the lambda groups don't reference removed (lambda) classes
* Assert that each method/class where a lambda is defined, references the correct lambda group instance that contains the implementation of the merged lambda

**Comment by heckej** [comment]
No test has been written for case 2: `Assert that each method/class where a lambda is defined, references the correct lambda group instance that contains the implementation of the merged lambda`.

**Comment by heckej** [comment]
After creating lambda groups for classes
```
IntrinsicMeasureBlocks$VerticalMaxWidth$1
IntrinsicMeasureBlocks$VerticalMaxWidth$1$1
IntrinsicMeasureBlocks$VerticalMaxWidth$1$2
```
it appears that not all inner lambda's have been merged (or at least not correctly). The reference to the `INSTANCE` field of the first inner lambda has correctly been replaced with an instantiation of the lambda group. In contrast, the reference to the `INSTANCE` field of the second inner lambda is still there:
```
return Integer.valueOf(RowColumnImplKt.access$intrinsicSize(list2, new LambdaGroup(2, 2), IntrinsicMeasureBlocks.HorizontalMaxWidth.1.2.INSTANCE, intValue3, intValue4, LayoutOrientation.Horizontal, LayoutOrientation.Horizontal));
```

It seems like the second inner lambda has been merged after the outer lambda, such that the code referencing the inner lambda was not updated inside the outer lambda's implementation in the lambda group.

**Comment by heckej** [comment]
When an inner class has been removed from an other class, the `u2classesCount` of the `InnerClassesAttribute` in the outer class is decremented by 1. As a result, the visitor that visits the `InnerClassesAttribute` will skip an iteration of the loop everytime one of the inner classes has been removed. 

This could be solved by:

* keeping track of the fact that some inner class has been removed, and revisiting the left-over inner classes
* removing the inner class attribute after the lambda group has been built

**Comment by heckej** [comment]
This issue has been fixed by commit [b7ba37489e0ca882b51f1e3f8730944260f1131c](https://github.com/heckej/proguard/commit/b7ba37489e0ca882b51f1e3f8730944260f1131c) in this repository and commit [ea25197d7d4ad6e19e788ac89bfccca047d10944](https://github.com/heckej/proguard-core/commit/ea25197d7d4ad6e19e788ac89bfccca047d10944) in [proguard-core](https://github.com/heckej/proguard-core).
